### PR TITLE
[Geo] Construct: Phase 2 - API Key Support

### PR DESCRIPTION
**Pull request by manavgurnani21**
## Problem

With the creation of three resources within the Amplify Geo vertical (`AmplifyMap`, `AmplifyPlace`, and `AmplifyCollection`), developers could provision instances of these resources. While `AmplifyCollection` provisions a Geofence Collection through CDK, the initial version of `AmplifyMap` and `AmplifyPlace` did not provision any resources. Their entire purpose was to perform access control over customer-defined access permissions over AWS-managed Location resources. However, customers can integrate these resources within their application front-end through API keys. This PR focuses primarily on the feature support and output integration of AWS Location API keys for these resources.

**Issue number, if available:**

## Changes

As part of the API key integration, there are three important segements to the implementation:

#### 1. API Key Provisioning

- Setup customer-end API key properties within `defineX()` (only one API key generated for Map/Place resource)
- Added `generateApiKey()` function for conditional provisioning
- Unit testing for API key provisioning

#### 2. Access Orchestration for API Keys

- Add API key definition to `GeoAccessBuilder` (maintains current DX of access orchestration)
- Conditional API key provisioning logic and key restriction mapping
- Additional error handling and unit testing for access orchestration

#### 3. Updating Amplify Outputs Schema
- Updating output schema within `backend-output-schemas/geo` to accomodate API keys
- Creating new amplify output schema to accomodate for geo API keys (`v1.4` -> `v1.5`)
- Bumping geo client configuration functionality to function with new schema
    - additional unit testing for new schema
    - new legacy conversion to `v0` schema (support for map styles deprecated)
    - sandbox command update (to run on `v1.5`)
- Additional unit testing for client config generation pipeline

### New DX for Geo Resources
``` typescript
export const generalMap = defineMap({
    name: 'testGeneralMap',
    access: (allow) => [
        allow.authenticated.to(['get']),
        allow.guest.to(['get']),
        allow.apiKey.to(['get']) // add API key access restrictions
    ],
    apiKeyProps: {
        apiKeyName: 'testKey'
    }
});

export const generalPlace = definePlace({
    name: 'testGeneralIndex',
    access: (allow) => [
        allow.authenticated.to(['search', 'geocode', 'autocomplete']),
        allow.guest.to(['search']),
        allow.apiKey.to(['geocode', 'search'])
    ],
    apiKeyProps: {
        apiKeyName: 'testIndexKey',
        noExpiry: true, // add API key proprerties
    }
});
```

**Corresponding docs PR, if applicable:**

## Validation

- 126 unit tests with >95% code coverage within `backend-geo`
- Additional tests added into `client-config` package pass tests
- Successful ESLint verification
- Successful API generation
- Successful deployment & amplify output validation on Sandbox project testing with above DX

### Unit Test Coverage
- Additional customer-end `defineX()` API testing
- Additional access orchestration testing
- API key generation testing (empty vs. non-empty)
- Error and edge-case handling tests for API keys
- Amplify backend output schema testing
- Amplify client configuration versioned testing

## Checklist

- [X] If this PR includes a functional change to the runtime behavior of the code, I have added or updated automated test coverage for this change.
- [ ] If this PR requires a change to the [Project Architecture README](../PROJECT_ARCHITECTURE.md), I have included that update in this PR.
- [ ] If this PR requires a docs update, I have linked to that docs PR above.
- [X] If this PR modifies E2E tests, makes changes to resource provisioning, or makes SDK calls, I have run the PR checks with the `run-e2e` label set.

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: .changeset/fluffy-icons-wave.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+'@aws-amplify/ai-constructs': patch
+'@aws-amplify/platform-core': patch
+'@aws-amplify/backend-auth': patch
+'@aws-amplify/backend-data': patch
+'@aws-amplify/plugin-types': patch
+'@aws-amplify/backend-geo': patch
+'@aws-amplify/backend-ai': patch
+'@aws-amplify/backend': patch
+'@aws-amplify/backend-cli': patch
+---
```
Comment (rtpascual) [inline .changeset/fluffy-icons-wave.md:18]: There should be a comment for every changeset on what changed here, in this case upgrading `aws-cdk-lib` peer dependencies.

File: packages/backend-geo/API.md [hunk 2]
```diff
@@ -37,7 +43,9 @@ export type AmplifyMapFactoryProps = Omit<AmplifyMapProps, 'outputStorageStrateg
 // @public (undocumented)
 export type AmplifyMapProps = {
     name: string;
+    isDefault?: boolean;
```
Comment (rtpascual) [inline packages/backend-geo/API.md]: I thought we can only define one map and place? Is there still a need for `isDefault` in these props?

File: packages/backend-geo/src/geo_access_policy_factory.test.ts [hunk 3]
```diff
@@ -495,4 +495,134 @@ void describe('GeoAccessPolicyFactory', () => {
...
+        'geo-places:SearchText',
+        'geo-places:Suggest',
+      ]);
+    });
+
+    void it('generateKeyActions returns empty array for empty input', () => {
+      const keyActions = geoAccessPolicyFactory.generateKeyActions([]);
+
+      assert.deepStrictEqual(keyActions, []);
+    });
```
Comment (rtpascual) [inline packages/backend-geo/src/geo_access_policy_factory.test.ts]: I'm wondering if we should allow this or throw to force users to add an action?

File: packages/client-config/src/client-config-schema/client_config_v1.5.ts [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+     */
+    maps?: {
+      /**
+       * @minItems 1
+       */
+      items?: {
+        name?: string;
+        key?: string;
+        [k: string]: unknown;
+      }[];
```
Comment (rtpascual) [inline packages/client-config/src/client-config-schema/client_config_v1.5.ts]: Looks like we dropped `[k: string]: AmazonLocationServiceConfig;` here, is there a reason we are making this typing more loose?

File: packages/client-config/src/client-config-writer/client_config_to_legacy_converter.ts [hunk 5]
```diff
@@ -23,9 +23,9 @@ export class ClientConfigLegacyConverter {
    */
   convertToLegacyConfig = (clientConfig: ClientConfig): ClientConfigLegacy => {
     // We can only convert from V1.4 of ClientConfig. For everything else, throw
```
Comment (rtpascual) [inline packages/client-config/src/client-config-writer/client_config_to_legacy_converter.ts]:
```suggestion
    // We can only convert from V1.5 of ClientConfig. For everything else, throw
```

File: packages/client-config/src/client-config-schema/client_config_v1.5.ts [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+    /**
+     * Maps from Amazon Location Service
+     */
+    maps?: {
+      /**
+       * @minItems 1
+       */
+      items?: {
+        name?: string;
+        key?: string;
```
Comment (rtpascual) [inline packages/client-config/src/client-config-schema/client_config_v1.5.ts]: I believe we should rename to `api_key` like in the data part of the schema.

File: packages/client-config/src/client-config-schema/client_config_v1.5.ts (hunk 4 above)
Comment (manavgurnani21) [reply packages/client-config/src/client-config-schema/client_config_v1.5.ts]: The `AmazonLocationServiceConfig` involved indexed map styles for every legacy map that was provisioned. This was a type used in older schemas and was not needed for the new version of the items schema (which only required names and API keys). The extra `[k: string]: unknown` was likely a schema conversion error since I intended on dropping it. I could alternatively change the internal type of `AmazonLocationServiceConfig` for the new schema and use that internally, but that could conflict with legacy types with the same name.

File: packages/backend-geo/src/geo_access_policy_factory.test.ts (hunk 3 above)
Comment (manavgurnani21) [reply packages/backend-geo/src/geo_access_policy_factory.test.ts]: We could throw a warning/error for users that have empty actions since each of the `allow` statements are optional anyways, so the user should only be calling them when they have actions to assign to them. Would this be the ideal DX for these functions?

File: packages/backend-geo/API.md [hunk 7]
```diff
@@ -92,9 +104,37 @@ export type GeoActionBuilder = {
...
+    cfnResources: {
+        cfnAPIKey?: CfnAPIKey;
+    };
+};
+
+// @public (undocumented)
+export type ResourceOutputs = {
+    name: string;
+    key?: string;
+};
```
Comment (sobolk) [inline packages/backend-geo/API.md]: Does this have to be public?
I don't see this being used by anything public.

File: packages/backend-geo/src/geo_outputs_aspect.test.ts [hunk 8]
```diff
@@ -179,34 +263,82 @@ void describe('AmplifyGeoOutputsAspect', () => {
...
+            maps: JSON.stringify({
+              default: "testMapResource",
+              items: [{
+                name: "testMapResource",
+                apiKey: "TOKEN_STRING"
+              }]
+            })
+          }
+        }
+       */
```
Comment (sobolk) [inline packages/backend-geo/src/geo_outputs_aspect.test.ts]: commented out code should be removed.

File: packages/backend-geo/src/geo_access_policy_factory.test.ts (hunk 3 above)
Comment (rtpascual) [reply packages/backend-geo/src/geo_access_policy_factory.test.ts]: Yes I believe we should throw an error here, if users write `allow.x.to()` there is intent to grant x some permission.

**Review by sobolk — COMMENTED** [review]
Overall looks. Good.
Left couple of comments in addition to what @rtpascual already posted.

File: packages/client-config/src/client-config-schema/client_config_v1.5.ts (hunk 4 above)
Comment (rtpascual) [reply packages/client-config/src/client-config-schema/client_config_v1.5.ts]: I see, since this is not a major version bump of the schema I say we should add `key` to `AmazonLocationServiceConfig` as optional and keep `items` level as
https://github.com/aws-amplify/amplify-backend/blob/f8d2b02fd55705b1d81dcde528b685e163b5a031/packages/client-config/src/client-config-schema/client_config_v1.4.ts#L223-L228.

Since the goal is that customers can switch from CDK to Amplify Geo constructs and their frontend does not break.

Comment (manavgurnani21) [reply packages/client-config/src/client-config-schema/client_config_v1.5.ts]: This makes sense, the only other concern is that the actual hidden value of the key cannot be accessed through CDK, so the `ref` value from the API key construct resolves to the name of the API key itself. Users would then have to use the CLI or SDK to get this API key value and use it within their resources. The items object would contain a key pointing to the same value, which I now realize is redundant. I could change items to be of type `AmazonLocationServiceConfig` and add that directly (maintains item layer but different typing) or we could find another way to get the hidden value or another structure. Let me know your thoughts on the same.

File: .changeset/fluffy-icons-wave.md (hunk 1 above)
Comment (manavgurnani21) [reply .changeset/fluffy-icons-wave.md:18]: Done. PTAL.
https://github.com/aws-amplify/amplify-backend/blob/9d562af95b47dea765a0a04dc47cb9826aca2659/.changeset/fluffy-icons-wave.md?plain=1#L1-L20

File: packages/backend-geo/API.md (hunk 2 above)
Comment (manavgurnani21) [reply packages/backend-geo/API.md]: Removed. PTAL.
https://github.com/aws-amplify/amplify-backend/blob/9d562af95b47dea765a0a04dc47cb9826aca2659/packages/backend-geo/API.md?plain=1#L43-L49

File: packages/backend-geo/src/geo_outputs_aspect.test.ts (hunk 8 above)
Comment (manavgurnani21) [reply packages/backend-geo/src/geo_outputs_aspect.test.ts]: This was more of a note on structure to clarify where the TOKEN assertion was coming from since the value was loosely parsed. Would this be helpful or redundant?
https://github.com/aws-amplify/amplify-backend/blob/7566d1be63aa7151a7a628277c9f58dfcbc8dfa1/packages/backend-geo/src/geo_outputs_aspect.test.ts#L307-L319

File: packages/backend-geo/src/geo_access_policy_factory.test.ts (hunk 3 above)
Comment (manavgurnani21) [reply packages/backend-geo/src/geo_access_policy_factory.test.ts]: Done. PTAL.
https://github.com/aws-amplify/amplify-backend/blob/ae1e1881c8b60e6353e444a8d3f18cba73be8986/packages/backend-geo/src/geo_access_orchestrator.ts#L68-L75

File: packages/client-config/src/client-config-writer/client_config_to_legacy_converter.ts (hunk 5 above)
Comment (manavgurnani21) [reply packages/client-config/src/client-config-writer/client_config_to_legacy_converter.ts]: Done. PTAL.
https://github.com/aws-amplify/amplify-backend/blob/ae1e1881c8b60e6353e444a8d3f18cba73be8986/packages/client-config/src/client-config-writer/client_config_to_legacy_converter.ts#L24-L30

**Review by sobolk — APPROVED** [review]

File: packages/client-config/src/client-config-schema/client_config_v1.5.ts (hunk 4 above)
Comment (rtpascual) [reply packages/client-config/src/client-config-schema/client_config_v1.5.ts]: By hidden value of the key, do you mean during synthesis this is a CDK token? I believe there's no concern if we treat this key the same as the AppSync API key in data. Since it is added to stack outputs and also in `amplify_outputs.json`.

Comment (manavgurnani21) [reply packages/client-config/src/client-config-schema/client_config_v1.5.ts]: Changes made, PTAL. The change in the schema is now primarily for search indices as they include API keys as well, they have been made similar to the maps schema.
https://github.com/aws-amplify/amplify-backend/blob/8e594a3b430e2ea1ef1881a7555eb0221c365ec9/packages/client-config/src/client-config-schema/client_config_v1.5.ts#L226-L242

File: packages/client-config/src/client-config-schema/client_config_v1.5.ts (hunk 6 above)
Comment (manavgurnani21) [reply packages/client-config/src/client-config-schema/client_config_v1.5.ts]: Renamed to `api_key_name` since hidden values are not CDK-accessible.
https://github.com/aws-amplify/amplify-backend/blob/8e594a3b430e2ea1ef1881a7555eb0221c365ec9/packages/client-config/src/client-config-schema/client_config_v1.5.ts#L299-L309

File: packages/backend-geo/API.md (hunk 7 above)
Comment (manavgurnani21) [reply packages/backend-geo/API.md]: Removed.

File: packages/backend-geo/src/geo_outputs_aspect.ts [hunk 9]
```diff
@@ -6,10 +6,16 @@ import { AmplifyPlace } from './place_resource.js';
...
 import { GeoOutput, geoOutputKey } from '@aws-amplify/backend-output-schemas';
-import { GeoResourceType, ResourceOutputs } from './types.js';
+import { GeoResourceType } from './types.js';
 
 export type GeoResource = AmplifyMap | AmplifyPlace | AmplifyCollection;
 
+type ResourceOutputs = {
+  name?: string;
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  api_key_name?: string;
```
Comment (rtpascual) [inline packages/backend-geo/src/geo_outputs_aspect.ts]: Nit: we can keep naming conventions here, i.e. `apiKeyName`, and in the contributer when we write to `amplify_outputs.json` that's when we change the field name to `api_key_name`.

This way we can remove this eslint-disable and keep naming consistent for each "layer" from outputs -> client config

Comment (manavgurnani21) [reply packages/backend-geo/src/geo_outputs_aspect.ts]: Done. PTAL.
https://github.com/aws-amplify/amplify-backend/blob/6a843f15c52d96b0571cf1b0d1b580092b2c09da/packages/backend-geo/src/geo_outputs_aspect.ts#L12-L17
